### PR TITLE
Json codable no type assumption in custom fromJson

### LINKED ISSUE #59580: [json_codable] the object passed to fromJson is not necessarily a Map
Currently json_codable assumes that a Map is passed to the fromJson constructor when that constructor is defined by the user. 

This is not supported:

```dart

@JsonCodable()
class A {
  final CustomStringSerialization customStringSerializationField;
}

class CustomStringSerialization {
  final String a;

  CustomStringSerialization(this.a);

  String toJson() => a;

  factory CustomStringSerialization.fromJson(String a) =>
      CustomStringSerialization(a);
}

```

Use cases:

  - deserializing values to object, for example a 'EUR' value could be deserialized to a currency object with its format method
  - it will be necessary for enum support

### PR DESCRIPTION
**Pull request by cedvdb**
- fix https://github.com/dart-lang/sdk/issues/59580
- related: https://github.com/dart-lang/sdk/issues/55989

 > json_codable wrongly assumes fromJson receives a Map

This allows an user defined class to have a `fromJson` method with a parameter that is not a `Map`. Before this change the `@JsonCodable` macro would check if a field has a fromJson constructor and assume that the parameter type was a `Map`. This PR uses the type of the parameter instead.

Note: I have this weird behavior where I can only run the tests with the macros enabled once per branch, so I have to switch branches and run the tests on another.

### CODE REVIEW

**Comment by cedvdb** [comment]
@jakemac53 not sure if updates to this package are welcome yet or not. I've another down the drain for enums but I cannot make it work currently because the behavior is different during test than in the augmented code which is puzzling

**Comment by cedvdb** [comment]
It seems there is already a PR for this here : https://github.com/dart-lang/sdk/pull/56665

**Comment by natebosch** [comment]
Can a class also define how to handle `null`? Would `factory Foo.fromJson(Object? o)` or `factory Foo.fromJson(String? s)` work? Can we add a test if so?

**Comment by cedvdb** [comment]
> Can a class also define how to handle `null`? Would `factory Foo.fromJson(Object? o)` or `factory Foo.fromJson(String? s)` work? Can we add a test if so?

It does not currently handle that case no

File: pkg/json/test/json_codable_test.dart [hunk 1]
```diff
@@ -38,6 +43,11 @@ void main() {
       expect(a.listOfSerializableField.single.x, 1);
       expect(a.setOfSerializableField.single.x, 2);
       expect(a.mapOfSerializableField['c']!.x, 3);
+      expect(a.customStringSerializationField, CustomStringSerialization('hi'));
+      expect(a.customNullableStringSerialization,
```
Comment (cedvdb) [inline pkg/json/test/json_codable_test.dart:47]: @natebosch added requested test

**Comment by cedvdb** [comment]
@natebosch I added a test for String? o

**Comment by jakemac53** [comment]
This is released now 👍

**Pull request closed without merging**